core/src/main/scala/kafka/common/InterBrokerSendThread.scala

```diff
@@ -46,32 +55,30 @@ abstract class InterBrokerSendThread(name: String,
     val now = time.milliseconds()
     var pollTimeout = Long.MaxValue
 
+    generateRequests().foreach { request =>
+      val completionHandler = request.handler
+      unsentRequests.put(request.destination,
+        networkClient.newClientRequest(request.destination.idString(), request.request, now, true, completionHandler))
+    }
+
     try {
-      for (request: RequestAndCompletionHandler <- generateRequests()) {
-        val destination = Integer.toString(request.destination.id())
-        val completionHandler = request.handler
-        val clientRequest = networkClient.newClientRequest(destination,
-          request.request,
-          now,
-          true,
-          completionHandler)
-
-        if (networkClient.ready(request.destination, now)) {
-          networkClient.send(clientRequest, now)
-        } else {
-          val header = clientRequest.makeHeader(request.request.latestAllowedVersion)
-          val disconnectResponse: ClientResponse = new ClientResponse(header, completionHandler, destination,
-            now /* createdTimeMs */ , now /* receivedTimeMs */ , true /* disconnected */ , null /* versionMismatch */ ,
-            null /* responseBody */)
-
-          // poll timeout would be the minimum of connection delay if there are any dest yet to be reached;
-          // otherwise it is infinity
-          pollTimeout = Math.min(pollTimeout, networkClient.connectionDelay(request.destination, now))
-
-          completionHandler.onComplete(disconnectResponse)
+      if (unsentRequests.hasRequests) {
+        for (node <- unsentRequests.nodes) {
+          val requestIterator = unsentRequests.requestIterator(node)
+          while (requestIterator.hasNext) {
+            val request = requestIterator.next
+            if (networkClient.ready(node, now)) {
+              networkClient.send(request, now)
+              requestIterator.remove()
+              unsentRequests.clean()
+            } else
+              pollTimeout = Math.min(pollTimeout, networkClient.connectionDelay(node, now))
```

Comment:
Suggested before, but I thought we might be able to make this simple and use the connection delay if there are any unsent requests and `Long.MaxValue` otherwise.